### PR TITLE
feat: add metrics for the Get Client Device Auth Token API

### PR DESCRIPTION
**Pull request by sameerzuberi**
Created the GetClientDeviceAuthToken.Success and GetClientDeviceAuthToken.Failure metrics. When an auth token is successfully or unsuccessfully retrieved, an event is emitted and handled by incrementing its respective metric. Both the event and handler are specific to the Get Client Device Auth Token API.

This change is necessary to allow for tracking the operational status of the Get Client Device Auth Token API. The change was tested by writing unit tests which ensure that the metric is correctly incremented upon events being emitted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/clientdevices/auth/metrics/ClientDeviceAuthMetrics.java [hunk 1]
```diff
@@ -130,6 +136,26 @@ public List<Metric> collectMetrics() {
...
+        metric = Metric.builder()
+                .namespace(NAMESPACE)
+                .name(METRIC_GET_CLIENT_DEVICE_AUTH_TOKEN_FAILURE)
+                .unit(TelemetryUnit.Count)
+                .aggregation(TelemetryAggregation.Sum)
+                .value(getClientDeviceAuthTokenFailure.getAndSet(0L))
+                .timestamp(timestamp)
+                .build();
+        metricsList.add(metric);
+
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/metrics/ClientDeviceAuthMetrics.java:158]: Do we want to remove metrics without counts?

File: src/main/java/com/aws/greengrass/ipc/GetClientDeviceAuthTokenOperationHandler.java [hunk 2]
```diff
@@ -97,8 +102,12 @@ public GetClientDeviceAuthTokenResponse handleRequest(GetClientDeviceAuthTokenRe
                 String sessionId =
                         clientDevicesAuthServiceApi.getClientDeviceAuthToken(MQTT_CREDENTIAL_TYPE, credentialMap);
                 GetClientDeviceAuthTokenResponse response = new GetClientDeviceAuthTokenResponse();
+                domainEvents.emit(new GetClientDeviceAuthTokenEvent(GetClientDeviceAuthTokenEvent
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/ipc/GetClientDeviceAuthTokenOperationHandler.java:105]: Following the logic from above, we'd then want to emit this at a lower layer. Moquette actually isn't using IPC. It's calling the service API directly. Ideally this should be coming from the session domain

File: src/main/java/com/aws/greengrass/clientdevices/auth/api/GetClientDeviceAuthTokenEvent.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+/*
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/api/GetClientDeviceAuthTokenEvent.java:1]: This should probably belong in the session domain. Follow the directory structure we're using for the other domain events (take the term quite literally -- a domain event is an event that a software domain emits)

File: src/main/java/com/aws/greengrass/clientdevices/auth/metrics/handlers/GetClientDeviceAuthTokenMetricHandler.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
+/*
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/metrics/handlers/GetClientDeviceAuthTokenMetricHandler.java:1]: Similarly, this isn't listening for `GetClientDeviceAuthTokenEvents`, it's listening to session creation events

File: src/main/java/com/aws/greengrass/clientdevices/auth/metrics/ClientDeviceAuthMetrics.java (hunk 1 above)
Comment (sameerzuberi) [reply src/main/java/com/aws/greengrass/clientdevices/auth/metrics/ClientDeviceAuthMetrics.java:158]: I can check if a metric has a positive value before building it to begin with

**Comment by jbutler** [comment]
I'm not concerned with performance. Using a stream and filter below is
probably easier to read. However you want to do it

File: src/main/java/com/aws/greengrass/ipc/GetClientDeviceAuthTokenOperationHandler.java (hunk 2 above)
Comment (jbutler) [reply src/main/java/com/aws/greengrass/ipc/GetClientDeviceAuthTokenOperationHandler.java:105]: https://github.com/aws-greengrass/aws-greengrass-client-device-auth/blob/8682bcb81b23022fe43a95f18fe67229d8acbad8/src/main/java/com/aws/greengrass/clientdevices/auth/session/SessionManager.java#L66

You could catch it, emit the metric, then rethrow

File: src/main/java/com/aws/greengrass/clientdevices/auth/metrics/ClientDeviceAuthMetrics.java [hunk 5]
```diff
@@ -130,6 +137,30 @@ public List<Metric> collectMetrics() {
...
+        metricsList = metricsList.stream()
```
Comment (jcosentino11) [inline src/main/java/com/aws/greengrass/clientdevices/auth/metrics/ClientDeviceAuthMetrics.java:160]: lets add a test for this

File: src/main/java/com/aws/greengrass/clientdevices/auth/session/handlers/SessionCreationHandler.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+package com.aws.greengrass.clientdevices.auth.session.handlers;
+
+import com.aws.greengrass.clientdevices.auth.api.DomainEvents;
+import com.aws.greengrass.clientdevices.auth.metrics.ClientDeviceAuthMetrics;
+import com.aws.greengrass.clientdevices.auth.session.events.SessionCreationEvent;
+
+import java.util.function.Consumer;
+import javax.inject.Inject;
+
+public class SessionCreationHandler implements Consumer<SessionCreationEvent> {
```
Comment (vaibhavmurkute) [inline src/main/java/com/aws/greengrass/clientdevices/auth/session/handlers/SessionCreationHandler.java:15]: Consider renaming this handler to reflect that its an _event handler_ and not Session-Creation handler (we actually have another class for that `SessionCreator`). 

Same goes for other _event_ handlers:
- `VerifyClientDeviceIdentityHandler`
- `CertificateSubscriptionHandler`

**Review by jcosentino11 — DISMISSED** [review]

File: src/main/java/com/aws/greengrass/clientdevices/auth/api/ClientDevicesAuthServiceApi.java [hunk 7]
```diff
@@ -11,6 +11,7 @@
 import com.aws.greengrass.clientdevices.auth.exception.AuthenticationException;
 import com.aws.greengrass.clientdevices.auth.exception.AuthorizationException;
 import com.aws.greengrass.clientdevices.auth.exception.CertificateGenerationException;
+import com.aws.greengrass.clientdevices.auth.iot.events.VerifyClientDeviceIdentityEvent;
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/api/ClientDevicesAuthServiceApi.java:14]: ?? shouldn't this cause a pmd violation? is it being used?

File: src/main/java/com/aws/greengrass/clientdevices/auth/certificate/handlers/CertificateSubscriptionEventHandler.java [hunk 8]
```diff
@@ -3,7 +3,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-package com.aws.greengrass.clientdevices.auth.metrics.handlers;
+package com.aws.greengrass.clientdevices.auth.certificate.handlers;
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/certificate/handlers/CertificateSubscriptionEventHandler.java:6]: Sorry I think I confused you. The handler should still be part of the metrics domain. This is the thing that knows about metrics.

File: src/main/java/com/aws/greengrass/clientdevices/auth/iot/handlers/VerifyClientDeviceIdentityEventHandler.java [hunk 9]
```diff
@@ -3,16 +3,16 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-package com.aws.greengrass.clientdevices.auth.metrics.handlers;
+package com.aws.greengrass.clientdevices.auth.iot.handlers;
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/clientdevices/auth/iot/handlers/VerifyClientDeviceIdentityEventHandler.java:6]: Same thing. Can keep this in metrics

File: src/main/java/com/aws/greengrass/clientdevices/auth/api/ClientDevicesAuthServiceApi.java (hunk 7 above)
Comment (sameerzuberi) [reply src/main/java/com/aws/greengrass/clientdevices/auth/api/ClientDevicesAuthServiceApi.java:14]: I'm using it down in `verifyClientDeviceIdentity()` to emit the event

**Review by jbutler — APPROVED** [review]

**Review by jcosentino11 — APPROVED** [review]

**Pull request merged**